__tests__/commands/upgrade.js

```diff
@@ -104,6 +104,34 @@ test.concurrent('upgrades from fixed version to latest', (): Promise<void> => {
   });
 });
 
+test.concurrent('upgrades package not in registry', (): Promise<void> => {
+  const packages = ['yarn-test-git-repo', 'e2e-test-repo'];
+  return runUpgrade(packages, {}, 'package-not-in-registry', async (config): ?Promise<void> => {
+    const lockfile = explodeLockfile(await fs.readFile(path.join(config.cwd, 'yarn.lock')));
+    const gitRemote = 'https://github.com/juanca/e2e-test-repo';
+
+    assert(
+      lockfile.includes(`"yarn-test-git-repo@${gitRemote}#package-not-in-registry":`),
+      'Lockfile should point to the same yarn-test-git-repo branch.',
+    );
+
+    assert(
+      lockfile.includes(`  resolved "${gitRemote}#5c57959f3c55a6cd6004e3855ca59ba98bfff56c"`),
```

Comment:
This assertion is currently failing. Seems like there is a deeper issue with upgrading exotic packages. 

This [assertion was switched to be more correct](https://github.com/yarnpkg/yarn/pull/2780/commits/f9186622cde0686174abf5710a05a7f4c1e3e692):

> Replace assertion of non-existence with existence
> The assertion was incorrect and this particular assertion was testing nothing. The whole test itself still provides some coverage: packages not in registry does not blow up `yarn upgrade <package name>` command.